homeassistant/components/lcn/helpers.py

```diff
@@ -200,6 +246,109 @@ def import_lcn_config(lcn_config: ConfigType) -> list[ConfigType]:
+        identifiers = {(DOMAIN, generate_unique_id(config_entry.entry_id, address))}
+
+        if device_config[CONF_ADDRESS][2]:  # is group
+            device_model = f"LCN group (g{address[0]:03d}{address[1]:03d})"
+            sw_version = None
+        else:  # is module
+            hardware_type = device_config[CONF_HARDWARE_TYPE]
+            if hardware_type in pypck.lcn_defs.HARDWARE_DESCRIPTIONS:
+                hardware_name = pypck.lcn_defs.HARDWARE_DESCRIPTIONS[hardware_type]
+            else:
+                hardware_name = pypck.lcn_defs.HARDWARE_DESCRIPTIONS[-1]
+            device_model = f"{hardware_name} (m{address[0]:03d}{address[1]:03d})"
+            sw_version = f"{device_config[CONF_SOFTWARE_SERIAL]:06X}"
+
+        device_registry.async_get_or_create(
+            config_entry_id=config_entry.entry_id,
+            identifiers=identifiers,
+            via_device=host_identifiers,
+            manufacturer="Issendorff",
+            sw_version=sw_version,
+            name=device_name,
+            model=device_model,
+        )
+
+
+async def async_update_device_config(
+    device_connection: DeviceConnectionType, device_config: ConfigType
+) -> None:
+    """Fill missing values in device_config with infos from LCN bus."""
+    is_group = device_config[CONF_ADDRESS][2]
+
+    # fetch serial info if device is module
+    if not is_group:  # is module
+        await device_connection.serial_known
+        if device_config[CONF_HARDWARE_SERIAL] == -1:
+            device_config[CONF_HARDWARE_SERIAL] = device_connection.hardware_serial
+        if device_config[CONF_SOFTWARE_SERIAL] == -1:
+            device_config[CONF_SOFTWARE_SERIAL] = device_connection.software_serial
+        if device_config[CONF_HARDWARE_TYPE] == -1:
+            device_config[CONF_HARDWARE_TYPE] = device_connection.hardware_type.value
+
+    # fetch name if device is module
+    if device_config[CONF_NAME] != "":
+        return
+
+    device_name = ""
+    if not is_group:
+        device_name = await device_connection.request_name()
+    if is_group or device_name == "":
+        module_type = "Group" if is_group else "Module"
+        device_name = (
+            f"{module_type} "
+            f"{device_config[CONF_ADDRESS][0]:03d}/"
+            f"{device_config[CONF_ADDRESS][1]:03d}"
+        )
+    device_config[CONF_NAME] = device_name
+
+
+async def async_update_config_entry(
+    hass: HomeAssistant, config_entry: ConfigEntry
+) -> None:
+    """Fill missing values in config_entry with infos from LCN bus."""
+    coros = []
+    for device_config in config_entry.data[CONF_DEVICES]:
+        device_connection = get_device_connection(
+            hass, device_config[CONF_ADDRESS], config_entry
+        )
+        coros.append(async_update_device_config(device_connection, device_config))
+
+    await asyncio.gather(*coros)
+
+    # schedule config_entry for save
+    hass.config_entries.async_update_entry(config_entry)
```

Comment:
We're not allowed to update the config entry data in place. Please pass the new data as a parameter to this function.